File: lib/crypto/space_key.js
/**
 * Space Key — AES-256-GCM encryption for Mind Protocol spaces.
 *
 * Spaces are encrypted with a symmetric key. Content is stored as
 * self-contained ciphertext strings: "base64(iv):base64(tag):base64(ciphertext)".
 * IV is regenerated on every encrypt call — never reused.
 *
 * Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
 */

'use strict';

const crypto = require('crypto');

// --- Constants ---

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;   // 96-bit IV for GCM (NIST recommended)
const TAG_LENGTH = 16;  // 128-bit authentication tag
const KEY_LENGTH = 32;  // 256-bit key

// --- Functions ---

/**
 * Generate a cryptographically random space key.
 * @returns {{ key: Buffer }} 32-byte symmetric key
 */
function generateSpaceKey() {
  return { key: crypto.randomBytes(KEY_LENGTH) };
}

/**
 * Encrypt plaintext content with a space key (AES-256-GCM).
 *
 * Output format: "base64(iv):base64(tag):base64(ciphertext)"
 * A fresh IV is generated on every call — never reused.
 *
 * @param {string} plaintext - The content to encrypt
 * @param {Buffer} key - 32-byte space key
 * @returns {string} Colon-separated base64 ciphertext string
 */
function encryptContent(plaintext, key) {
  if (!Buffer.isBuffer(key) || key.length !== KEY_LENGTH) {
    throw new Error(`Space key must be a ${KEY_LENGTH}-byte Buffer`);
  }
  if (typeof plaintext !== 'string') {
    throw new Error('Plaintext must be a string');
  }

  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv, { authTagLength: TAG_LENGTH });

  const encrypted = Buffer.concat([
    cipher.update(plaintext, 'utf8'),
    cipher.final(),
  ]);

  const tag = cipher.getAuthTag();

  return [
    iv.toString('base64'),
    tag.toString('base64'),
    encrypted.toString('base64'),
  ].join(':');
}

/**
 * Decrypt a ciphertext string produced by encryptContent.
 *
 * @param {string} encrypted - "base64(iv):base64(tag):base64(ciphertext)"
 * @param {Buffer} key - 32-byte space key
 * @returns {string} Decrypted plaintext
 */
function decryptContent(encrypted, key) {
  if (!Buffer.isBuffer(key) || key.length !== KEY_LENGTH) {
    throw new Error(`Space key must be a ${KEY_LENGTH}-byte Buffer`);
  }
  if (typeof encrypted !== 'string') {
    throw new Error('Encrypted content must be a string');
  }

  const parts = encrypted.split(':');
  if (parts.length !== 3) {
    throw new Error('Invalid ciphertext format — expected "iv:tag:ciphertext"');
  }

  const iv = Buffer.from(parts[0], 'base64');
  const tag = Buffer.from(parts[1], 'base64');
  const ciphertext = Buffer.from(parts[2], 'base64');

  if (iv.length !== IV_LENGTH) {
    throw new Error(`Invalid IV length: expected ${IV_LENGTH}, got ${iv.length}`);
  }
  if (tag.length !== TAG_LENGTH) {
    throw new Error(`Invalid tag length: expected ${TAG_LENGTH}, got ${tag.length}`);
  }

  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv, { authTagLength: TAG_LENGTH });
  decipher.setAuthTag(tag);

  const decrypted = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final(),
  ]);

  return decrypted.toString('utf8');
}

/**
 * Detect whether a string looks like encrypted content (the colon-separated base64 format).
 *
 * @param {string} content - String to test
 * @returns {boolean} true if the string matches the ciphertext format
 */
function isEncrypted(content) {
  if (typeof content !== 'string') return false;
  // Three base64 segments separated by colons
  // Base64 charset: A-Z a-z 0-9 + / =
  const BASE64 = '[A-Za-z0-9+/]+=*';
  const pattern = new RegExp(`^${BASE64}:${BASE64}:${BASE64}$`);
  return pattern.test(content);
}

// --- Exports ---

module.exports = {
  ALGORITHM,
  IV_LENGTH,
  TAG_LENGTH,
  KEY_LENGTH,
  generateSpaceKey,
  encryptContent,
  decryptContent,
  isEncrypted,
};

File: python/crypto/space_key.py
"""
Space key encryption for Mind Protocol.

AES-256-GCM encryption using the `cryptography` package.
Ciphertext format: "base64(iv):base64(tag):base64(ciphertext)"
Cross-compatible with the JS implementation.

Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
"""

import os
import re
import base64

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

IV_LENGTH = 12   # 96-bit nonce for AES-GCM
TAG_LENGTH = 16  # 128-bit authentication tag
KEY_LENGTH = 32  # 256-bit key


def generate_space_key() -> bytes:
    """Generate a random 256-bit (32-byte) AES key for space encryption.

    Returns:
        32 random bytes suitable for AES-256-GCM.
    """
    return os.urandom(KEY_LENGTH)


def encrypt_content(plaintext: str, key: bytes) -> str:
    """Encrypt plaintext using AES-256-GCM.

    A fresh IV is generated for every call, ensuring unique ciphertexts
    even for identical inputs.

    Args:
        plaintext: The string to encrypt.
        key: A 32-byte AES-256 key.

    Returns:
        Ciphertext in the format "base64(iv):base64(tag):base64(ciphertext)".

    Raises:
        ValueError: If the key length is not 32 bytes.
    """
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Key must be {KEY_LENGTH} bytes, got {len(key)}")

    iv = os.urandom(IV_LENGTH)
    aesgcm = AESGCM(key)

    # AESGCM.encrypt appends the 16-byte tag to the ciphertext
    ct_with_tag = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)

    # Split: last TAG_LENGTH bytes are the tag, the rest is ciphertext
    ciphertext = ct_with_tag[:-TAG_LENGTH]
    tag = ct_with_tag[-TAG_LENGTH:]

    iv_b64 = base64.b64encode(iv).decode("ascii")
    tag_b64 = base64.b64encode(tag).decode("ascii")
    ct_b64 = base64.b64encode(ciphertext).decode("ascii")

    return f"{iv_b64}:{tag_b64}:{ct_b64}"


def decrypt_content(encrypted: str, key: bytes) -> str:
    """Decrypt an AES-256-GCM ciphertext string.

    Args:
        encrypted: Ciphertext in the format "base64(iv):base64(tag):base64(ciphertext)".
        key: The 32-byte AES-256 key used for encryption.

    Returns:
        The original plaintext string.

    Raises:
        ValueError: If the key length is wrong or the ciphertext format is invalid.
        cryptography.exceptions.InvalidTag: If decryption fails (wrong key or tampered data).
    """
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Key must be {KEY_LENGTH} bytes, got {len(key)}")

    parts = encrypted.split(":")
    if len(parts) != 3:
        raise ValueError(
            f"Invalid ciphertext format: expected 3 colon-separated parts, got {len(parts)}"
        )

    iv = base64.b64decode(parts[0])
    tag = base64.b64decode(parts[1])
    ciphertext = base64.b64decode(parts[2])

    if len(iv) != IV_LENGTH:
        raise ValueError(f"IV must be {IV_LENGTH} bytes, got {len(iv)}")
    if len(tag) != TAG_LENGTH:
        raise ValueError(f"Tag must be {TAG_LENGTH} bytes, got {len(tag)}")

    aesgcm = AESGCM(key)

    # Reconstruct the format AESGCM.decrypt expects: ciphertext + tag
    plaintext_bytes = aesgcm.decrypt(iv, ciphertext + tag, None)
    return plaintext_bytes.decode("utf-8")


def is_encrypted(content: str) -> bool:
    """Detect whether a string looks like encrypted content.

    Matches the JS implementation: checks for exactly three colon-separated
    base64 segments (charset A-Za-z0-9+/=, at least one char each).
    Does NOT validate decoded lengths -- this is a format check, not a
    cryptographic validation.

    Args:
        content: The string to test.

    Returns:
        True if the string matches the ciphertext format, False otherwise.
    """
    if not isinstance(content, str):
        return False

    # Same regex logic as JS: three base64 segments separated by colons
    base64_segment = r"[A-Za-z0-9+/]+=*"
    pattern = rf"^{base64_segment}:{base64_segment}:{base64_segment}$"
    return bool(re.match(pattern, content))
